### ISSUE TITLE
Potential changes to DisposeBag

### ISSUE THREAD

**Issue by hmlongco** [ReactiveX/RxSwift#1880]
Consider the following code in a main viewController:

```swift
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        coordinator
            .edit(element: viewModel.element(at: indexPath.row))
            .subscribe(onNext: { [weak self] element in
                self?.viewModel.update(indexPath.row, element))
            })
            .disposed(by: disposeBag)
        tableView.deselectRow(at: indexPath, animated: true)
    }
```

Here the coordinator launches an edit screen and returns an observable for the entered value. Every time the cell is tapped, a new disposable is added to the disposeBag.

If a value is entered and the user hits save, the subscription will return a new value and complete. If the user just cancels, the subscription will also complete as in that case the observable is set to complete without returning a value. 

In both cases, the subscription will properly be disposed.

However, over time the disposeBag disposable array will gradually fill up with empty "disposed" disposables. The above is just an example, as there are several ways in which a disposeBag's disposables can complete or error out and leave the bag holding a disposed disposable entry.

I believe that there's also a situation (case 2 below) where a subscription is made on an observable that happens to be a "just" value. In that case the subscription immediately fires the just value, completes, and the entire subscription disposes before the disposable even has a chance to be inserted into the disposeBag. (There are also several ways this can occur, and not just with just.)

Hence I was wondering if the following change to DisposeBag might be in order:

```swift
    private func _insert(_ disposable: Disposable) -> Disposable? {
        _lock.lock(); defer { _lock.unlock() }
        if _isDisposed {
            return disposable
        }

        // case 2
        if disposable.isDisposed {
            return nil
        }

        /// case 1
        _disposables = _disposables.compactMap { $0.isDisposed ? nil : $0 }
        _disposables.append(disposable)

        return nil
    }
```
If case 2 we don't bother inserting a disposed disposable.

In case 1, we garbage collect and compact the disposable array, removing completed, disposed disposables. 

An alternative would be to provide a "CompactingDisposeBag" that subclasses DisposeBag with the above change such that people who might need the above behavior could use it when and if desired.

Note above code was simplified by a small protocol extension to Disposable:

```swift
public extension Disposable {
    var isDisposed: Bool { return false }
}
```

Practically every real disposable is Cancelable, so the extension is overwritten and it defaults to a safe mode should there be user-written Disposables that don't conform to Cancelable.

**Comment by kzaher** [comment]
Hi @hmlongco ,

If you are doing
```
override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        coordinator
            .edit(element: viewModel.element(at: indexPath.row))
            .subscribe(onNext: { [weak self] element in
                self?.viewModel.update(indexPath.row, element))
            })
            .disposed(by: disposeBag)
        tableView.deselectRow(at: indexPath, animated: true)
    }
```
then I would argue that you are using it in non-idiomatic way.

Having said that, if ` .edit(element: viewModel.element(at: indexPath.row))` completes, than you don't need to use dispose bag at all. All resources will be cleaned up once that observable completes.

You can also do something like this.
```
func viewDidLoad() {
  tableView.rx.itemSelected
         .flatMapLatest { indexPath in
              coordinator
                   .edit(element: viewModel.element(at: indexPath.row))
                  .map { $0 in ($0, indexPath) }
           }
           .subscribe(onNext: { [weak self] ... in
                self?.viewModel.update(indexPath.row, element))
            })
            .disposed(by: disposeBag)
}
```

**Comment by hmlongco** [comment]
Sorry, was out and just got back. To respond, yes, that usage could be considered non-idiomatic. That said, you know as well as I do that there are a lot of people out there that still use RxSwift in a non-idiomatic, imperative way.

The following code was from a very early RxSwift app...
```
    func save() {
        guard let request = buildRequest() else {
            return
        }
        processing.accept(true)
        updater.update(request)
            .subscribe(onNext: { _ in
                self.processing.accept(false)
                self.updated.onNext(self.successMessage)
            }, onError: { (error) in
                self.processing.accept(false)
                self.error.accept(error as? APIError)
            })
            .disposed(by: disposeBag)
    }
```
Where it suffers from the primary issue shown above. Is it "proper" RxSwift? No. Is it a way for someone to integrate RxSwift into an app? Absolutely.

**Comment by hmlongco** [comment]
Also note that there were two issues mentioned, the second being that it doesn't make a lot of sense to put an already disposed disposable into the disposebag.  There are many ways in which a subscription can complete before the result of the subscription function is inserted.

**Comment by kzaher** [comment]
@hmlongco 

I think it would be hard to win this argument with the following code :)
```swift
updater.update(request)
            .subscribe(onNext: { _ in
                self.processing.accept(false)
                self.updated.onNext(self.successMessage)
            }, onError: { (error) in
                self.processing.accept(false)
                self.error.accept(error as? APIError)
            })
            .disposed(by: disposeBag)
```

There are two cases to consider.
* `updater.update(request)` doesn't complete -> It is a memory leak because it is referencing `self` and dispose bag will never be disposed.
* `updater.update(request)` completes -> then putting the stuff in dispose bag will just increase the amount of disposables until VC deallocates.

In any case it's a bad piece of code. 

I am not saying that it is not possible to design a better API from our end and that we are not bear a part of the blame, but designing an API is not easy. I can't estimate easily how people will abuse our API until I see it, and I don't think that changing the `DisposeBag` in this way would help people as much as you think.

**Comment by hmlongco** [comment]
@kzaher Didn't say it was wonderful, just said it was an early attempt at integrating RxSwift. ;)

The API updater.update(request) would always complete or error, so no leak (good). However, every update would add a new disposable (bad.

Unfortunately, most of the RxSwift "tutorials" out there emphasize always putting the disposable into the bag,  which in this case leads to the ever increasing bag of disposed disposables. (Not to mention the compiler complains if you don't use the result somewhere.)

The following code is a somewhat better approach using Single, but the let _ = is a bit of black magic unless one is a little more familiar with RxSwift's internal functionality and the fact that the subscription persists past the scope of the enclosing function. (Something we initially thought required the bag.)

```
let _ = updater.update(request)
            .subscribe(onSuccess: { [weak self] _ in
                self?.processing.accept(false)
                self?.updated.onNext(self.successMessage)
            }, onError: { [weak self] (error) in
                self?.processing.accept(false)
                self?.error.accept(error as? APIError)
            })
```
As you say, I'm not sure how many people the change might help, but I am sure that it's possible to "leak" disposed disposables in the bag array.

I'd thought about keeping a counter and only compressing the bag after every 20 or 30 or so insertions, but didn't want to overly complicate the sample.

At any rate, that's why I dropped it here first as a "potential" change so we could discuss it... and we did. Thanks for listening.
